### PR TITLE
fix: Fix encode Snowflake ID for object identifiers

### LINKED ISSUE #2058: Cannot create a snowflake_user with provider 0.70.1
**Provider Version**

0.70.1
terraform-provider-snowflake_v0.70.1

**Terraform Version**

Terraform v1.1.9

**Describe the bug**

Creating a new snowflake user resource fails on apply with a plugin did not respond error.

Example plan:
```
# module.account.snowflake_user.user will be created
  + resource "snowflake_user" "user" {
      + default_role       = "ROLE"
      + default_warehouse  = "WAREHOUSE"
      + disabled           = false
      + display_name       = (sensitive value)
      + email              = (sensitive value)
      + first_name         = (sensitive value)
      + has_rsa_public_key = (known after apply)
      + id                 = (known after apply)
      + last_name          = (sensitive value)
      + login_name         = (known after apply)
      + name               = (sensitive value)
    }

Plan: 1 to add, 0 to change, 0 to destroy.
```

After applying
```
Error: Plugin did not respond
│
│   with module.account.snowflake_user.user,
│   on ../../../modules/account/users.tf line 1344, in resource "snowflake_user" "user":
│ 1344: resource "snowflake_user" "user" {
│
│ The plugin encountered an error, and failed to respond to the
│ plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.

Stack trace from the terraform-provider-snowflake_v0.70.1 plugin:

panic: interface conversion: sdk.ObjectIdentifier is sdk.DatabaseObjectIdentifier, not sdk.AccountObjectIdentifier

goroutine 85 [running]:
github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/resources.ReadUser(0xc00050d600, {0x1ac67c0?, 0xc000b40000?})
	github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/resources/user.go:244 +0x96a
github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/resources.CreateUser(0x0?, {0x1ac67c0?, 0xc000b40000?})
	github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/resources/user.go:236 +0xbc8
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0x1de2900?, {0x1de2900?, 0xc000e7b7d0?}, 0xd?, {0x1ac67c0?, 0xc000b40000?})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.26.1/helper/schema/resource.go:695 +0x178
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc00025cc40, {0x1de2900, 0xc000e7b7d0}, 0xc00087cdd0, 0xc00050d480, {0x1ac67c0, 0xc000b40000})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.26.1/helper/schema/resource.go:837 +0xa85
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc000277920, {0x1de2900?, 0xc000e7b6b0?}, 0xc00019a640)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.26.1/helper/schema/grpc_provider.go:1021 +0xe8d
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc00085c640, {0x1de2900?, 0xc000e7acc0?}, 0xc000aabe30)
	github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/tf5server/server.go:818 +0x574
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x1a79800?, 0xc00085c640}, {0x1de2900, 0xc000e7acc0}, 0xc000aabdc0, 0x0)
	github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:385 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000768960, {0x1df22c0, 0xc0008e0000}, 0xc000526fc0, 0xc0005afe30, 0x2a3c3e0, 0x0)
	google.golang.org/grpc@v1.53.0/server.go:1336 +0xd33
google.golang.org/grpc.(*Server).handleStream(0xc000768960, {0x1df22c0, 0xc0008e0000}, 0xc000526fc0, 0x0)
	google.golang.org/grpc@v1.53.0/server.go:1704 +0xa36
google.golang.org/grpc.(*Server).serveStreams.func1.2()
	google.golang.org/grpc@v1.53.0/server.go:965 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
	google.golang.org/grpc@v1.53.0/server.go:963 +0x28a

Error: The terraform-provider-snowflake_v0.70.1 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.

Operation failed: failed running terraform apply (exit 1)
```

**Expected behavior**

Expected user to be created in Snowflake without an error - sometimes the user is created sometimes it is not.

**Code samples and commands**

Example TF code:
```
resource "snowflake_user" "user" {
  name     = "USER"
  email    = "USER@EMAIL.COM"
  disabled = false

  first_name   = "Test"
  last_name    = "User"
  display_name = "Test User"

  default_warehouse = "WAREHOUSE"
  default_role      = "ROLE"
}
```

### PR DESCRIPTION
**Pull request by sfc-gh-asawicki**
Fix encoding SnowflakeIDs for SDK object identifiers.

It still needs work for other types; we will revisit this with identifiers rework.

Fix #2058

### CODE REVIEW

File: pkg/helpers/helpers.go [hunk 1]
```diff
@@ -65,10 +65,18 @@ func EncodeSnowflakeID(attributes ...interface{}) string {
 	// is attribute already an object identifier?
 	if len(attributes) == 1 {
 		if id, ok := attributes[0].(sdk.ObjectIdentifier); ok {
-			// remove quotes and replace dots with pipes
-			parts := strings.Split(id.FullyQualifiedName(), ".")
-			for i, part := range parts {
-				parts[i] = strings.Trim(part, `"`)
+			parts := make([]string, 0)
+			switch v := id.(type) {
```
Comment (sfc-gh-jcieslak) [inline pkg/helpers/helpers.go:69]: Should we also consider cases where a pointer to an identifier is passed? If not then, we'll hit the default branch

File: pkg/helpers/helpers_test.go [hunk 2]
```diff
@@ -71,3 +72,83 @@ func TestDecodeSnowflakeParameterID(t *testing.T) {
...
+func TestEncodeSnowflakeID(t *testing.T) {
+	testCases := map[string]struct {
+		identifier        sdk.ObjectIdentifier
+		expectedEncodedID string
+	}{
+		"encodes account object identifier": {
+			identifier:        sdk.NewAccountObjectIdentifier("database"),
+			expectedEncodedID: `database`,
+		},
+		"encodes quoted account object identifier": {
```
[the remark this answers is not in the document]
Comment (sfc-gh-asawicki) [reply pkg/helpers/helpers_test.go:86]: We can address it in identifiers rework. For now, it works differently. When we use SDK, we create a new object identifier that first trims the quotes and later is always used as a fully qualified name.

[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/helpers/helpers_test.go:86]: Yeah, let's move the discussion on it when we'll be tackling identifier rework

File: pkg/helpers/helpers.go (hunk 1 above)
Comment (sfc-gh-asawicki) [reply pkg/helpers/helpers.go:69]: I will add this.

Comment (sfc-gh-asawicki) [reply pkg/helpers/helpers.go:72]: Done.

Comment (sfc-gh-jcieslak) [reply pkg/helpers/helpers.go:72]: I think the matching cases could be joined but other than that 👍

File: pkg/helpers/helpers_test.go [hunk 3]
```diff
@@ -76,7 +77,7 @@ func TestDecodeSnowflakeParameterID(t *testing.T) {
 // TODO: add tests for non object identifiers
 func TestEncodeSnowflakeID(t *testing.T) {
 	testCases := map[string]struct {
-		identifier        sdk.ObjectIdentifier
+		identifier        any
```
Comment (sfc-gh-jcieslak) [inline pkg/helpers/helpers_test.go:80]: sdk.ObjectIdentifier can hold both values and pointers to values, but that's a nitpick

**Review by sfc-gh-jcieslak — APPROVED** [review]

File: pkg/helpers/helpers.go (hunk 1 above)
Comment (sfc-gh-asawicki) [reply pkg/helpers/helpers.go:72]: Tried that, and it worked only for AccountObjectIdentifier :(

File: pkg/helpers/helpers_test.go (hunk 3 above)
Comment (sfc-gh-asawicki) [reply pkg/helpers/helpers_test.go:80]: Done

**Review by sfc-gh-jcieslak — APPROVED** [review]

**Pull request merged**